src/lib/statistics/cardinality_estimator.cpp

```diff
@@ -83,6 +86,86 @@ std::optional<float> estimate_null_value_ratio_of_column(const TableStatistics&
   return std::nullopt;
 }
 
+std::shared_ptr<TableStatistics> prune_column_statistics(
+    const std::shared_ptr<TableStatistics>& table_statistics, const std::vector<ColumnID>& pruned_column_ids,
+    const std::vector<std::shared_ptr<AbstractExpression>>& output_expressions,
+    std::optional<ExpressionUnorderedSet>& required_columns) {
+  if (!required_columns && pruned_column_ids.empty()) {
+    return table_statistics;
+  }
+
+  // Prune `pruned_column_ids` and all statistics for unused columns from the statistics.
+  auto output_column_statistics = std::vector<std::shared_ptr<const BaseAttributeStatistics>>(
+      table_statistics->column_statistics.size() - pruned_column_ids.size());
+
+  auto pruned_column_ids_iter = pruned_column_ids.begin();
+
+  for (auto input_column_id = ColumnID{0}, output_column_id = ColumnID{0};
+       input_column_id < table_statistics->column_statistics.size(); ++input_column_id) {
+    // Skip `stored_column_id` if it is in the sorted vector `pruned_column_ids`.
+    if (pruned_column_ids_iter != pruned_column_ids.end() && input_column_id == *pruned_column_ids_iter) {
+      ++pruned_column_ids_iter;
+      continue;
+    }
+
+    // Create a dummy statistics object if the column is not actually used for estimations.
```

Comment:
Minor, but aligning the if-else with the comment makes it easier to read. So switch if/else and maybe use `required columns && !req->contains)`  ?